tstime/tstime.go

```diff
@@ -6,117 +6,251 @@
 package tstime
 
 import (
-	"errors"
-	"fmt"
-	"strings"
-	"sync"
+	"context"
+	"math/bits"
+	"sync/atomic"
 	"time"
 )
 
-// zoneOf returns the RFC3339 zone suffix (either "Z" or like
-// "+08:30"), or the empty string if it's invalid or not something we
-// want to cache.
-func zoneOf(s string) string {
-	if strings.HasSuffix(s, "Z") {
-		return "Z"
-	}
-	if len(s) < len("2020-04-05T15:56:00+08:00") {
-		// Too short, invalid? Let time.Parse fail on it.
-		return ""
-	}
-	zone := s[len(s)-len("+08:00"):]
-	if c := zone[0]; c == '+' || c == '-' {
-		min := zone[len("+08:"):]
-		switch min {
-		case "00", "15", "30":
-			return zone
-		}
-	}
-	return ""
+// Time is an interface to all things that deal with the current time,
+// and the passage of time. It's meant to be used as a replacement for
+// the functions of the same name in the `time` and `context`
+// packages, so that different implementations of time can be swapped
+// in.
+type Time interface {
+	// After is like time.After.
+	After(time.Duration) <-chan time.Time
+	// Sleep is like time.Sleep.
+	Sleep(time.Duration)
+	// Tick is like time.Tick. TODO(danderson) remove? It leaks
+	// goroutines and requires shutting up the linter, and we probably
+	// shouldn't use it?...
+	Tick(time.Duration) <-chan time.Time
+
+	// Since is like time.Since.
+	Since(time.Time) time.Duration
+	// Until is like time.Until.
+	Until(time.Time) time.Duration
+
+	// NewTicker is like time.NewTicker.
+	NewTicker(time.Duration) *time.Ticker
+
+	// Now is like time.Now
+	Now() time.Time
+
+	// AfterFunc is like time.AfterFunc.
+	AfterFunc(time.Duration, func()) *Timer
+	// NewTimer is like time.NewTimer.
+	NewTimer(time.Duration) *Timer
+
+	// WithTimeout is like context.WithTimeout.
+	WithTimeout(context.Context, time.Duration) (context.Context, context.CancelFunc)
+	// WithDeadline is like context.WithDeadline.
+	WithDeadline(context.Context, time.Time) (context.Context, context.CancelFunc)
```

Comment:
these will read oddly at callsites, depending on how you name the interface value's variable:

    time.WithDeadline ?
    time.WithTimeout?

Doesn't say anything about context.

Maybe add "Context" prefix to both of these? That gross?